src/transformers/trainer.py

```diff
@@ -2255,17 +2255,21 @@ def _inner_training_loop(
                     # Optimizer step
                     self.optimizer.step()
                     optimizer_was_run = not self.accelerator.optimizer_step_was_skipped
+                    # Store the current LR before stepping
+                    learning_rate = self._get_learning_rate()
                     if optimizer_was_run:
-                        # Delay optimizer scheduling until metrics are generated
+                        # Delay optimizer scheduling until metrics are generated and logged
                         if not isinstance(self.lr_scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau):
                             self.lr_scheduler.step()
-
                     model.zero_grad()
                     self.state.global_step += 1
                     self.state.epoch = epoch + (step + 1 + steps_skipped) / steps_in_epoch
                     self.control = self.callback_handler.on_step_end(args, self.state, self.control)
 
-                    self._maybe_log_save_evaluate(tr_loss, grad_norm, model, trial, epoch, ignore_keys_for_eval)
+                    self._maybe_log_save_evaluate(
+                        tr_loss, grad_norm, learning_rate, model, trial, epoch, ignore_keys_for_eval
+                    )
```

Comment:
Why not move this line above L2260? It seems it's logging a few values before there's any change to the global state e.g. `epoch`

Reply:
That changes any behavior if users want to modify something with `on_step_end` and fully adjusts when that callback is called. Seems far too reaching for this IMO :) We want that to be a thing

Reply:
(And if we do this, it will make some tests fail IIRC because reported vs expected changes wildly)

Reply:
If I've understood correctly, this means that the elements being logged aren't all in sync i.e. some will be reporting values before a step and some after

Reply:
Yes, which I think for the scheduler is important because stepping it only reports what the next LR will be when grabbing, which is not what we want. We want the applied scheduler. For the rest not as much I **think**, I'll look back through some eval reports to make sure that it is just an isolated incident.

Reply:
Sorry, I don't think I've fully understood what should be happening. 

> Yes, which I think for the scheduler is important because stepping it only reports what the next LR will be when grabbing, which is not what we want. We want the applied scheduler. 

Three questions: 
* In terms of what we "want" - do you mean what we want to pass to `maybe_log_save_evaluate` here, or something else?
* What is the difference between the scheduler and the applied scheduler? 
* Why can't we use the applied scheduler

I think it would be good to have @pacman100 review this first, as he's more familiar with all of these objects

> For the rest not as much I think, I'll look back through some eval reports to make sure that it is just an isolated incident.

Were you able to review reports and confirm behaviour is as expected?